### PR TITLE
Dev/aamstutz/add instances

### PR DESCRIPTION
**Pull request by amstuta**
# Description

Add resource and datasource for instance(s)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [x] Test A: `make testacc TESTARGS="-run TestAccCloudProjectInstance_basic"`
- [x] Test B: `make testacc TESTARGS="-run TestAccDataSourceCloudProjectInstances_basic"`
- [x] Test C: `make testacc TESTARGS="-run TestAccDataSourceCloudProjectInstance_basic"`

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings or issues
- [x] I have added acceptance tests that prove my fix is effective or that my feature works
- [x] New and existing acceptance tests pass locally with my changes
- [x] I ran succesfully `go mod vendor` if I added or modify `go.mod` file

### CODE REVIEW

File: ovh/data_cloud_project_instances.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+										Description: "IP version",
+										Computed:    true,
+									},
+								},
+							},
+						},
+						"attached_volumes": {
+							Type:        schema.TypeList,
+							Computed:    true,
+							Description: " Volumes attached to the instance",
```
Comment (deathiop) [inline ovh/data_cloud_project_instances.go:58]:
```suggestion
							Description: "Volumes attached to the instance",
```

File: ovh/data_cloud_project_instances.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+							Computed:    true,
+						},
+						"flavor_name": {
+							Type:        schema.TypeString,
+							Description: "Flavor name",
+							Computed:    true,
+						},
+						"name": {
+							Type:        schema.TypeString,
+							Description: "Flavor name",
```
Comment (deathiop) [inline ovh/data_cloud_project_instances.go:81]:
```suggestion
							Description: "Instance name",
```

File: ovh/data_cloud_project_instances.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+						},
+						"attached_volumes": {
+							Type:        schema.TypeList,
+							Computed:    true,
+							Description: " Volumes attached to the instance",
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"id": {
+										Type:        schema.TypeString,
+										Description: "Volume Id",
```
Comment (deathiop) [inline ovh/data_cloud_project_instances.go:63]:
```suggestion
										Description: "Volume id",
```

File: ovh/data_cloud_project_instance.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+			},
+			"attached_volumes": {
+				Type:        schema.TypeSet,
+				Computed:    true,
+				Description: "Volumes attached to the instance",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:        schema.TypeString,
+							Description: "Volume Id",
```
Comment (deathiop) [inline ovh/data_cloud_project_instance.go:60]:
```suggestion
							Description: "Volume id",
```

File: ovh/data_cloud_project_instance.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+				Computed:    true,
+			},
+			"image_id": {
+				Type:        schema.TypeString,
+				Description: "Image id",
+				Computed:    true,
+			},
+			"ssh_key": {
+				Type:        schema.TypeString,
+				Description: "Instance task state",
```
Comment (deathiop) [inline ovh/data_cloud_project_instance.go:93]: Nope

File: ovh/data_cloud_project_instances.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+							Computed:    true,
+						},
+						"image_id": {
+							Type:        schema.TypeString,
+							Description: "Image id",
+							Computed:    true,
+						},
+						"ssh_key": {
+							Type:        schema.TypeString,
+							Description: "Instance task state",
```
Comment (deathiop) [inline ovh/data_cloud_project_instances.go:96]: Nope

File: ovh/data_cloud_project_instances.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+	log.Printf("[DEBUG] Will read instances from project %s in region %s", serviceName, region)
+	if err := config.OVHClient.Get(endpoint, &res); err != nil {
+		return helpers.CheckDeleted(d, err, endpoint)
+	}
+
+	instances := make([]map[string]interface{}, len(res))
+	ids := make([]string, len(res))
+	for i, instance := range res {
+		instances[i] = instance.ToMap()
+		ids = append(ids, instance.Id)
```
Comment (deathiop) [inline ovh/data_cloud_project_instances.go:131]:
```suggestion
		ids[i] = instance.Id
```

File: ovh/resource_cloud_project_instance.go [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+						},
+					},
+				},
+			},
+			"ssh_key_create": {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				ForceNew:    true,
+				MaxItems:    1,
+				Description: "Creatting SSH Keypair",
```
Comment (deathiop) [inline ovh/resource_cloud_project_instance.go:153]:
```suggestion
				Description: "Creating SSH Keypair",
```

File: ovh/resource_cloud_project_instance.go [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+			},
+			"attached_volumes": {
+				Type:        schema.TypeSet,
+				Computed:    true,
+				Description: " Volumes attached to the instance",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:        schema.TypeString,
+							Description: "Volume Id",
```
Comment (deathiop) [inline ovh/resource_cloud_project_instance.go:219]:
```suggestion
							Description: "Volume id",
```

File: ovh/resource_cloud_project_instance.go [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+							Description: "IP version",
+							Computed:    true,
+						},
+					},
+				},
+			},
+			"attached_volumes": {
+				Type:        schema.TypeSet,
+				Computed:    true,
+				Description: " Volumes attached to the instance",
```
Comment (deathiop) [inline ovh/resource_cloud_project_instance.go:214]:
```suggestion
				Description: "Volumes attached to the instance",
```

File: ovh/resource_cloud_project_instance.go [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+	r := &CloudProjectInstanceResponse{}
+
+	endpoint := fmt.Sprintf("/cloud/project/%s/region/%s/instance/%s",
+		url.PathEscape(serviceName),
+		url.PathEscape(region),
+		url.PathEscape(id),
+	)
+
+	if err := config.OVHClient.Get(endpoint, r); err != nil {
+		return diag.Errorf("Error calling post %s:\n\t %q", endpoint, err)
```
Comment (deathiop) [inline ovh/resource_cloud_project_instance.go:333]:
```suggestion
		return diag.Errorf("Error calling get %s:\n\t %q", endpoint, err)
```

File: website/docs/d/cloud_project_instance.markdown [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `service_name` - (Required, Forces new resource) The id of the public cloud project. If omitted,
+  the `OVH_CLOUD_PROJECT_SERVICE` environment variable is used.
+* `region` - (Required, Forces new resource) Instance region.
+* `instance_id` - (Required, Forces new resource) Instance id
```
Comment (deathiop) [inline website/docs/d/cloud_project_instance.markdown:28]:
```suggestion
* `instance_id` - (Required, Forces new resource) Instance id.
```

File: website/docs/d/cloud_project_instance.markdown [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `addresses` - Instance IP addresses
+  * `ip` - IP address
+  * `version` - IP version
+* `attached_volumes` - Volumes attached to the instance
+  * `id` - Volume Id
```
Comment (deathiop) [inline website/docs/d/cloud_project_instance.markdown:38]:
```suggestion
  * `id` - Volume id
```

File: website/docs/d/cloud_project_instances.markdown [hunk 14]
```diff
@@ -0,0 +1,43 @@
+---
+subcategory : "Cloud Project"
+---
+
+# ovh_cloud_project_instance
+**This datasource uses a Beta API**
+
+Use this data source to get the list of instance in a region of a public cloud project.
```
Comment (deathiop) [inline website/docs/d/cloud_project_instances.markdown:8]:
```suggestion
Use this data source to get the list of instances in a region of a public cloud project.
```

**Review by deathiop — CHANGES_REQUESTED** [review]

File: ovh/data_cloud_project_instance.go [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+
+func dataSourceCloudProjectInstance() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceCloudProjectInstanceRead,
+		Schema: map[string]*schema.Schema{
+			"service_name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				DefaultFunc: schema.EnvDefaultFunc("OVH_CLOUD_PROJECT_SERVICE", nil),
+				Description: "Service name of the resource representing the id of the cloud project.",
```
Comment (deathiop) [inline ovh/data_cloud_project_instance.go:20]:
```suggestion
				Description: "Service name of the resource representing the id of the cloud project",
```

File: ovh/data_cloud_project_instance.go [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+				Computed:    true,
+			},
+			"image_id": {
+				Type:        schema.TypeString,
+				Description: "Image id",
+				Computed:    true,
+			},
+			"ssh_key": {
+				Type:        schema.TypeString,
+				Description: "SSH Keypair name",
```
Comment (deathiop) [inline ovh/data_cloud_project_instance.go:93]:
```suggestion
				Description: "SSH Key pair name",
```

File: website/docs/r/cloud_project_instance.markdown [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+* `boot_from` - (Required, Forces new resource) Boot the instance from an image or a volume
+  * `image_id` - (Mandatory only if volume_id not used, Forces new resource) Instance image id
+  * `volume_id` - (Mandatory only if image_id not used, Forces new resource) Instance volume id
+* `group`- (Optional, Forces new resource) Start instance in group
+  * `group_id` - (Optional, Forces new resource) Group id
+* `name` - (Required, Forces new resource) Instance name
+* `ssh_key` - (Mandatory only if ssh_key_create not used, Forces new resource) Existing SSH Keypair
+  * `name` - (Optional, Forces new resource) SSH Keypair name
+* `ssh_key_create` - (Mandatory only if ssh_key not used, Forces new resource) Unix cron pattern
+  * `name` - (Optional, Forces new resource) SSH Keypair name
```
Comment (rbeuque74) [inline website/docs/r/cloud_project_instance.markdown:56]:
```suggestion
  * `name` - (Optional, Forces new resource) SSH Key pair name
```

**Review by rbeuque74 — COMMENTED** [review]
Add a check of flavor availability in the tests to make sure they are not failing if the stock of this flavor is zero

File: ovh/resource_cloud_project_instance.go [hunk 18]
```diff
@@ -0,0 +1,345 @@
...
+				Type:        schema.TypeSet,
+				Optional:    true,
+				ForceNew:    true,
+				MaxItems:    1,
+				Description: "Existing SSH Keypair",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:        schema.TypeString,
+							Description: "SSH Keypair name",
```
Comment (deathiop) [inline ovh/resource_cloud_project_instance.go:139]:
```suggestion
							Description: "SSH Key pair name",
```

if you repush

File: ovh/resource_cloud_project_instance.go [hunk 19]
```diff
@@ -0,0 +1,345 @@
...
+				Description: "Instance name",
+				Required:    true,
+				ForceNew:    true,
+			},
+			"ssh_key": {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				ForceNew:    true,
+				MaxItems:    1,
+				Description: "Existing SSH Keypair",
```
Comment (deathiop) [inline ovh/resource_cloud_project_instance.go:134]:
```suggestion
				Description: "Existing SSH Key pair",
```

File: ovh/data_cloud_project_instances.go [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+
+	log.Printf("[DEBUG] Will read instances from project %s in region %s", serviceName, region)
+	if err := config.OVHClient.Get(endpoint, &res); err != nil {
+		return helpers.CheckDeleted(d, err, endpoint)
+	}
+
+	instances := make([]map[string]interface{}, len(res))
+	ids := make([]string, 0, len(res))
+	for i, instance := range res {
+		instances[i] = instance.ToMap()
```
Comment (deathiop) [inline ovh/data_cloud_project_instances.go:130]: Switch to append

File: ovh/types_cloud_project_instance.go [hunk 21]
```diff
@@ -0,0 +1,244 @@
...
+	obj := make(map[string]interface{})
+	obj["flavor_id"] = v.FlavorId
+	obj["flavor_name"] = v.FlavorName
+	obj["image_id"] = v.ImageId
+	obj["id"] = v.Id
+	obj["name"] = v.Name
+	obj["ssh_key"] = v.SshKey
+	obj["task_state"] = v.TaskState
+
+	addresses := make([]map[string]interface{}, 0)
```
Comment (deathiop) [inline ovh/types_cloud_project_instance.go:85]:
```suggestion
	addresses := make([]map[string]interface{}, 0, len(v.Addresses))
```

File: ovh/types_cloud_project_instance.go [hunk 22]
```diff
@@ -0,0 +1,244 @@
...
+	for i := range v.Addresses {
+		address := make(map[string]interface{})
+		address["ip"] = v.Addresses[i].Ip
+		address["version"] = v.Addresses[i].Version
+		addresses = append(addresses, address)
+	}
+	obj["addresses"] = addresses
+
+	attachedVolumes := make([]map[string]interface{}, 0)
```
Comment (deathiop) [inline ovh/types_cloud_project_instance.go:94]: Same

File: ovh/types_cloud_project_instance.go [hunk 23]
```diff
@@ -0,0 +1,244 @@
...
+	networkSet := i.(*schema.Set).List()
+	for _, network := range networkSet {
+		mapping := network.(map[string]interface{})
+		networkOutput.Public = mapping["public"].(bool)
+	}
+	return &networkOutput
+}
+
+func (cpir *CloudProjectInstanceCreateOpts) FromResource(d *schema.ResourceData) {
+	cpir.Flavor = GetFlaorId(d.Get("flavor"))
```
Comment (deathiop) [inline ovh/types_cloud_project_instance.go:234]:
```suggestion
	cpir.Flavor = GetFlavorId(d.Get("flavor"))
```

File: ovh/types_cloud_project_instance.go [hunk 24]
```diff
@@ -0,0 +1,244 @@
...
+	obj["attached_volumes"] = a.Id
+	return obj
+}
+
+type CloudProjectInstanceResponseList struct {
+	Id   string `json:"id"`
+	Name string `json:"name"`
+}
+
+func GetFlaorId(i interface{}) *Flavor {
```
Comment (deathiop) [inline ovh/types_cloud_project_instance.go:124]:
```suggestion
func GetFlavorId(i interface{}) *Flavor {
```

**Review by deathiop — APPROVED** [review]

**Review by rbeuque74 — APPROVED** [review]
